src/levelset/levelSetSegmentationObject.cpp

```diff
@@ -2872,19 +2885,34 @@ std::array<double,3> LevelSetSegmentationObject::_evalGradient(const std::array<
 
     // Evaluate the distance of the point from the surface
     LevelSetSegmentationSurfaceInfo::SegmentConstIterator supportItr = getSurface().getCellConstIterator(support);
-    double distance = m_surfaceInfo->evalDistance(point, supportItr, signedLevelSet);
+
+    int nSegmentVertices = supportItr->getVertexCount();
+    BITPIT_CREATE_WORKSPACE(lambda, double, nSegmentVertices, ReferenceElementInfo::MAX_ELEM_VERTICES);
+    std::array<double, 3> projectionPoint = m_surfaceInfo->evalProjection(point, supportItr, lambda);
+    std::array<double, 3> gradient = point - projectionPoint;
+
+    double unsignedDistance = norm2(gradient);
 
     // Early return if the point lies on the surface
-    if (evalValueSign(distance) == 0) {
+    double distanceTolerance = m_surfaceInfo->getSurface().getTol();
+    if (utils::DoubleFloatingEqual()(unsignedDistance, 0., distanceTolerance, distanceTolerance)) {
         if (signedLevelSet) {
             return m_surfaceInfo->evalNormal(point, supportItr);
         } else {
             return {{0., 0., 0.}};
         }
     }
+    gradient /= unsignedDistance;
 
     // Evaluate levelset gradient
-    std::array<double,3> gradient = m_surfaceInfo->evalDistanceVector(point, supportItr) / distance;
+    if (signedLevelSet) {
+       std::array<double, 3> pseudoNormal = m_surfaceInfo->evalPseudoNormal(supportItr, lambda);
+       double normalGradient = dotProduct(gradient, pseudoNormal);
+       if (utils::DoubleFloatingEqual()(normalGradient, 0., distanceTolerance, distanceTolerance)) {
+            throw std::runtime_error("Unable to evaluate gradient: the point lies on the normal plane!");
+       }
+       gradient *= sign(normalGradient);
+    }
```

Comment:
Unless this optimization yields a significant speed-up, I don't like duplicating the code that evaluate the distance in the function that evaluates the gradient. Even if the speed-up is significant, the function may need some rework because it seems it will give wrong results if the point lies on the normal plane and signedLevelSet is false.

Reply:
I did the optimization in function evalDistance and I kept the initial version of the gradient evaluation function.